cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,104 @@
+
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.strimzi.operator.common.VertxUtil;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.QuorumInfo;
+
+import static java.lang.Math.ceil;
+
+/**
+ * Provides a method that determines whether it's safe to restart a KRaft controller
+ */
+class KafkaQuorumCheck {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaQuorumCheck.class.getName());
+    private final Reconciliation reconciliation;
+    private final Admin admin;
+    private final Vertx vertx;
+    private final long controllerQuorumFetchTimeoutMs;
+
+    KafkaQuorumCheck(Reconciliation reconciliation, Admin ac, Vertx vertx, long controllerQuorumFetchTimeoutMs) {
+        this.reconciliation = reconciliation;
+        this.admin = ac;
+        this.vertx = vertx;
+        this.controllerQuorumFetchTimeoutMs = controllerQuorumFetchTimeoutMs;
+    }
+
+    /**
+     * Returns future that completes with true if the given controller can be rolled based on the quorum state. Quorum is considered
+     * healthy if the majority of controllers have caught up with the quorum leader within the controller.quorum.fetch.timeout.ms.
+     */
+    Future<Boolean> canRollController(int podId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller {} can be rolled", podId);
+        return describeMetadataQuorum().map(info -> {
+            boolean canRoll = isQuorumHealthyWithoutPod(podId, info);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Restart pod {} would affect the quorum", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the quorum leader id");
+        return describeMetadataQuorum().map(QuorumInfo::leaderId).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean isQuorumHealthyWithoutPod(int podId, QuorumInfo info) {
+        int leaderId = info.leaderId();
+        if (leaderId < 0) {
+            LOGGER.warnCr(reconciliation, "No quorum leader is found because the leader id is set to {}", leaderId);
+            return false;
+        }
+        Map<Integer, Long> replicaStates = info.voters().stream().collect(Collectors.toMap(
```

Comment:
I find the terminology at bit hard to follow. We store the state of `voters` into a variable called `replicaStates`. The number of entries is then called `totalNumOfControllers`. It seems we use voters, replicas and controllers to refer to the same thing. Should we settle on a single term?

Reply:
How about we stick to term  `controller` everywhere? So `replicaStates`  becomes `controllerStates`?